### PR TITLE
CPM-1065: Update Attribute entity

### PR DESCRIPTION
**Pull request by Beninho**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [X] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Structure/Bundle/Doctrine/ORM/Repository/AttributeRepository.php [hunk 1]
```diff
@@ -365,4 +376,34 @@ public function findAttributesByFamily(FamilyInterface $family)
 
         return $qb->getQuery()->getResult();
     }
+
+    public function updateMainIdentifier(AttributeInterface $attribute): void
```
Comment (Beninho) [inline src/Akeneo/Pim/Structure/Bundle/Doctrine/ORM/Repository/AttributeRepository.php:380]: public ?

File: tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php [hunk 2]
```diff
@@ -180,23 +180,29 @@ public function getAttributeIdsUseableInGrid($codes = null, $groupIds = null)
...
+            return  $attribute->getType() === AttributeTypes::IDENTIFIER && $attribute->isMainIdentifier();
         })->first();
 
+        if (false === $attribute) {
+            $attribute = $this->attributes->filter(function (AttributeInterface $attribute): bool {
+                return  $attribute->getType() === AttributeTypes::IDENTIFIER && 'sku' === $attribute->getCode();
+            })->first();
+        }
+
         return false !== $attribute ? $attribute : null;
```
Comment (Beninho) [inline tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php:195]: throw exception if null

File: src/Akeneo/Pim/Structure/Component/Repository/AttributeRepositoryInterface.php [hunk 3]
```diff
@@ -77,18 +77,14 @@ public function getAttributeIdsUseableInGrid($codes = null, $groupIds = null);
...
-    public function getIdentifier();
+    public function getIdentifier(): ?AttributeInterface;
 
     /**
      * Get the identifier code
-     *
-     * @return string
      */
-    public function getIdentifierCode();
+    public function getIdentifierCode(): ?string;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Component/Repository/AttributeRepositoryInterface.php:87]:
```suggestion
    public function getIdentifierCode(): string;
```

File: src/Akeneo/Pim/Structure/Component/Repository/AttributeRepositoryInterface.php [hunk 4]
```diff
@@ -146,4 +142,6 @@ public function findAttributesByFamily(FamilyInterface $family);
      * @return array
      */
     public function findAvailableAxes($locale);
+
+    public function updateMainIdentifier(AttributeInterface $attribute): void;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Component/Repository/AttributeRepositoryInterface.php:146]: to move that in another PR (when we allow multiple identifier attributes)

File: src/Akeneo/Pim/Structure/Component/Validator/Constraints/SingleIdentifierAttributeValidator.php [hunk 5]
```diff
@@ -43,10 +43,10 @@ public function validate($attribute, Constraint $constraint)
             throw new UnexpectedTypeException($constraint, SingleIdentifierAttribute::class);
         }
 
-        if (AttributeTypes::IDENTIFIER === $attribute->getType()) {
+        if (AttributeTypes::IDENTIFIER === $attribute->getType() && null !== $this->attributeRepository->getIdentifier()) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Component/Validator/Constraints/SingleIdentifierAttributeValidator.php:46]: no need to update this class

File: tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php [hunk 6]
```diff
@@ -180,23 +180,33 @@ public function getAttributeIdsUseableInGrid($codes = null, $groupIds = null)
     /**
      * {@inheritdoc}
      */
-    public function getIdentifier()
+    public function getIdentifier(): ?AttributeInterface
```
Comment (mmetayer) [inline tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php:183]:
```suggestion
    public function getIdentifier(): AttributeInterface
```

File: tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php [hunk 7]
```diff
@@ -180,23 +180,33 @@ public function getAttributeIdsUseableInGrid($codes = null, $groupIds = null)
...
+        }
+
+        return $attribute;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function getIdentifierCode()
+    public function getIdentifierCode(): ?string
```
Comment (mmetayer) [inline tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php:205]:
```suggestion
    public function getIdentifierCode(): string
```

File: tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php [hunk 8]
```diff
@@ -260,4 +270,9 @@ public function findAvailableAxes($locale)
     {
         throw new NotImplementedException(__METHOD__);
     }
+
+    public function updateMainIdentifier(AttributeInterface $attribute): void
+    {
+        throw new NotImplementedException(__METHOD__);
+    }
```
Comment (mmetayer) [inline tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php:277]:
```suggestion

```

File: tests/back/Acceptance/spec/Attribute/InMemoryAttributeRepositorySpec.php [hunk 9]
```diff
@@ -107,10 +107,11 @@ function it_throws_an_exception_if_saved_object_is_not_an_attribute(\StdClass $o
 
     function it_gets_the_identifier_attribute()
     {
-        $identifier = (new Attribute())->setType(AttributeTypes::IDENTIFIER);
+        $identifier = (new Attribute())
+            ->setType(AttributeTypes::IDENTIFIER)
+            ->setCode('sku');
 
         $this->save($identifier);
-        $this->save((new Attribute())->setCode('name'));
```
Comment (mmetayer) [inline tests/back/Acceptance/spec/Attribute/InMemoryAttributeRepositorySpec.php:113]: why remove this line?

File: src/Akeneo/Pim/Structure/Bundle/MainIdentifier/ChangeMainIdentifierHandler.php [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class ChangeMainIdentifierHandler
+{
+    public function __construct(
+        private readonly AttributeRepositoryInterface $attributeRepository,
+    ) {
+    }
+
+    public function __invoke(ChangeMainIdentifier $changeMainIdentifier): void
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/MainIdentifier/ChangeMainIdentifierHandler.php:24]: we need tests for this handler

Comment (Balamung) [reply src/Akeneo/Pim/Structure/Bundle/MainIdentifier/ChangeMainIdentifierHandler.php:24]: Do we need to keep the handler in this PR ? We can add it back when we also implement `updateMainIdentifier` in the repo

File: src/Akeneo/Pim/Structure/Bundle/EventSubscriber/Install/InitMainIdentifierSubscriber.php [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            InstallerEvents::POST_DB_CREATE => 'initMainIdentifier'
+        ];
+    }
+
+    public function initMainIdentifier(GenericEvent $event): void
+    {
+        if (!$this->mainIdentifierColumnExists()) {
```
Comment (Balamung) [inline src/Akeneo/Pim/Structure/Bundle/EventSubscriber/Install/InitMainIdentifierSubscriber.php:38]: Add todo

File: src/Akeneo/Pim/Structure/Bundle/EventSubscriber/Install/InitMainIdentifierSubscriber.php [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    ) {
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            InstallerEvents::POST_DB_CREATE => 'initMainIdentifier'
```
Comment (Balamung) [inline src/Akeneo/Pim/Structure/Bundle/EventSubscriber/Install/InitMainIdentifierSubscriber.php:32]: POST FIXTURE LOAD

Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/EventSubscriber/Install/InitMainIdentifierSubscriber.php:32]: I'd be more comfortable if the suscriber listsned to the `InstallerEvents::POST_LOAD_FIXTURE` event

File: tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php [hunk 13]
```diff
@@ -180,23 +180,47 @@ public function getAttributeIdsUseableInGrid($codes = null, $groupIds = null)
...
     {
         $attribute = $this->attributes->filter(function (AttributeInterface $attribute): bool {
-            return  $attribute->getType() === AttributeTypes::IDENTIFIER;
+            return  $attribute->getType() === AttributeTypes::IDENTIFIER && $attribute->isMainIdentifier();
         })->first();
 
-        return false !== $attribute ? $attribute : null;
+        if (false === $attribute) {
+            $attribute = $this->attributes->filter(function (AttributeInterface $attribute): bool {
+                return  $attribute->getType() === AttributeTypes::IDENTIFIER && 'sku' === $attribute->getCode();
```
Comment (mmetayer) [inline tests/back/Acceptance/Attribute/InMemoryAttributeRepository.php:199]:
```suggestion
                return  $attribute->getType() === AttributeTypes::IDENTIFIE;
```

**Review by mmetayer — APPROVED** [review]
